codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -170,22 +241,30 @@ public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf dat
             Http2LocalFlowController flowController = flowController();
             int bytesToReturn = data.readableBytes() + padding;
 
-            boolean shouldIgnore = true;
+            final boolean shouldIgnore;
             try {
                 shouldIgnore = shouldIgnoreHeadersOrDataFrame(ctx, streamId, stream, "DATA");
-            } finally {
-                if (shouldIgnore) {
-                    // Ignoring this frame. We still need to count the frame towards the connection flow control
-                    // window, but we immediately mark all bytes as consumed.
-                    flowController.receiveFlowControlledFrame(stream, data, padding, endOfStream);
-                    flowController.consumeBytes(stream, bytesToReturn);
+            } catch (Http2Exception e) {
+                // Ignoring this frame. We still need to count the frame towards the connection flow control
+                // window, but we immediately mark all bytes as consumed.
+                flowController.receiveFlowControlledFrame(stream, data, padding, endOfStream);
+                flowController.consumeBytes(stream, bytesToReturn);
+                throw e;
+            } catch (Throwable t) {
+                throw connectionError(INTERNAL_ERROR, t, "Unhandled error on data stream id %d", streamId);
+            }
 
-                    // Verify that the stream may have existed after we apply flow control.
-                    verifyStreamMayHaveExisted(streamId);
+            if (shouldIgnore) {
+                // Ignoring this frame. We still need to count the frame towards the connection flow control
+                // window, but we immediately mark all bytes as consumed.
+                flowController.receiveFlowControlledFrame(stream, data, padding, endOfStream);
```

Comment:
Yes. This PR was doen while resolving issues from the h2spec application. This change is described in the commit message:

```
DefaultHttp2ConnectionDecoder#onData calls shouldIgnoreHeadersOrDataFrame but may swallow exceptions from this method. This means a STREAM_RST frame may not be sent when it should for an unknown stream and thus violate the RFC. The exception is no longer swallowed.
```